dicomutil/dicomutil.go

```diff
@@ -31,24 +37,51 @@ func main() {
 	if data == nil {
 		log.Panic("Error reading %s: %v", path, err)
 	}
-	log.Printf("Error reading %s: %v", path, err)
 	if *printMetadata {
 		for _, elem := range data.Elements {
 			fmt.Printf("%v\n", elem.String())
 		}
 	}
 	if *extractImages {
-		n := 0
 		for _, elem := range data.Elements {
 			if elem.Tag == dicomtag.PixelData {
 				data := elem.Value[0].(dicom.PixelDataInfo)
-				for _, frame := range data.Frames {
-					path := fmt.Sprintf("image.%d.jpg", n) // TODO: figure out the image format
-					n++
-					ioutil.WriteFile(path, frame, 0644)
-					fmt.Printf("%s: %d bytes\n", path, len(frame))
+				if data.Encapsulated {
+					var wg sync.WaitGroup
+					for frameIndex, frame := range data.EncapsulatedFrames {
+						wg.Add(1)
+						go generateEncapsulatedImage(frame, frameIndex, &wg)
+					}
+					wg.Wait()
+				} else {
+					var wg sync.WaitGroup
+					for frameIndex, frame := range data.NativeFrames {
+						wg.Add(1)
+						go generateNativeImage(frame, frameIndex, &wg) // parse image as gray scale
+					}
+					wg.Wait()
 				}
 			}
 		}
 	}
+	log.Println("Complete.")
+}
+
+func generateEncapsulatedImage(frame []byte, frameIndex int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	path := fmt.Sprintf("image_%d.jpg", frameIndex) // TODO: figure out the image format
+	ioutil.WriteFile(path, frame, 0644)
+	fmt.Printf("%s: %d bytes\n", path, len(frame))
+}
+
+func generateNativeImage(frame [][]int, frameIndex int, wg *sync.WaitGroup) {
+	defer wg.Done()
+	i := image.NewGray16(image.Rect(0, 0, 512, 512)) //TODO(suyash): dont assume 512 x 512
+	for j := 0; j < len(frame); j++ {
+		i.SetGray16(j%512, j/512, color.Gray16{Y: uint16(frame[j][0])}) // for now, assume we're not overflowing uint16, assume gray image
```

Comment:
Right now the utility tool only supports outputting gray images. Color output will be added in future diffs